Classes/Private/Manager/Audio/OCTAudioEngine.m

```diff
@@ -0,0 +1,237 @@
+@property (nonatomic, assign) AUNode ioNode;
+@property (nonatomic, assign) AudioUnit ioUnit;
+@property (nonatomic, assign) AudioComponentDescription ioUnitDescription;
+
+@end
+
+@implementation OCTAudioEngine
+
+#pragma mark - LifeCycle
+-(instancetype)init
+{
+    self = [super init];
+    if (!self)
+        return nil;
+
+    _ioUnitDescription.componentType = kAudioUnitType_Output;
+    _ioUnitDescription.componentSubType = kAudioUnitSubType_VoiceProcessingIO;
+    _ioUnitDescription.componentManufacturer = kAudioUnitManufacturer_Apple;
+    _ioUnitDescription.componentFlags = 0;
+    _ioUnitDescription.componentFlagsMask = 0;
+
+    NewAUGraph(&_processingGraph);
+    NewAUGraph(&_processingGraph);
+
+    AUGraphAddNode(_processingGraph,
+                   &_ioUnitDescription,
+                   &_ioNode);
+
+    AUGraphOpen(_processingGraph);
+
+    AUGraphNodeInfo(_processingGraph, _ioNode, NULL, &_ioUnit);
+    return self;
+}
+
+#pragma mark - Audio Controls
+-(BOOL)startAudioFlow:(NSError **)error
+{
+    if (![self startAudioSession:error] ||
+        ![self microphoneInput:YES error:error] ||
+        ![self setUpStreamFormat:error] ||
+        ![self initializeGraph:error] ||
+        ![self startGraph:error]) {
+        return NO;
+    }
+
+    return YES;
+}
+
+-(BOOL)stopAudioFlow:(NSError **)error
+{
+    OSStatus status = AUGraphStop(_processingGraph);
+    if (status != noErr) {
+        *error = [self createErrorWithCode:status
+                               description:@"AUGraphStop"
+                             failureReason:@"Failed to stop Graph"];
+        return NO;
+    }
+
+    status = AUGraphUninitialize(_processingGraph);
+    if (status != noErr) {
+        *error = [self createErrorWithCode:status
+                               description:@"AUGraphUninitialize"
+                             failureReason:@"Failed to unintialize graph"];
+        return NO;
+    }
+
+    AVAudioSession *session = [AVAudioSession sharedInstance];
+    return [session setActive:NO error:error];
+}
+
+-(BOOL)microphoneInput:(BOOL)enable error:(NSError **)error
+{
+    UInt32 enableInput = (enable)? 1 : 0;
+    OSStatus status = AudioUnitSetProperty(
+                                           _ioUnit,
+                                           kAudioOutputUnitProperty_EnableIO,
+                                           kAudioUnitScope_Input,
+                                           kInputBus,
+                                           &enableInput,
+                                           sizeof (enableInput)
+                                           );
+    if (status != noErr) {
+        *error = [self createErrorWithCode:status
+                               description:@"Microphone Enable/Disable"
+                             failureReason:@"Unable to disable/enable Mic Input"];
+        return NO;
+    }
+
+    return YES;
+}
+
+-(BOOL)outputEnable:(BOOL)enable error:(NSError **)error
```

Comment:
This method and `microphoneInput` looks almost same except for `kAudioUnitScope_Output`/`kAudioUnitScope_Input` parameter. I think you can move this code to a method and call it from here
```
[self changeScope:kAudioUnitScope_Output enable:enable error:error];
```